contributors/design-proposals/core-metrics-pipeline.md

```diff
@@ -0,0 +1,315 @@
+
+Metrics requirements, based on kubernetes component needs, are as follows:
+ - Kubelet
+  - Node-level capacity and availability metrics for Disk and Memory
+  - Pod-level usage metrics for Disk and Memory
+ - Scheduler
+  - Node-level capacity and availability metrics for Disk, CPU, and Memory
+  - Pod-level usage metrics for Disk, CPU, and Memory
+  - Container-level usage metrics for Disk, CPU, and Memory
+ - Horizontal-Pod-Autoscaler (HPA)
+  - Node-level capacity and availability metrics for CPU and Memory
+  - Pod-level usage metrics for CPU and Memory
+
+More details on how I intend to achieve these high level goals can be found in the Implementation Plan.
+
+In order to continue to provide the full summary API, either the kubelet or a stand-alone version of cAdvisor will need to publish these metrics.
+
+This Core Metrics API will be versioned to account for version-skew between kubernetes components.
+
+This proposal purposefully omits many metrics that may eventually become core metrics.  This is by design.  Once metrics are needed for an internal use case, they can be added to the core metrics API.
+
+### Proposed Core Metrics API:
+
+An important difference between the current summary api and the proposed core metrics api is that per-pod stats in the core metrics api contain only usage data, and not capacity-related statistics.  This is more accurate since a pod's resource capacity is really defined by its "requests" and "limits", and it is a better reflection of how the kubelet uses the data.  The kubelet finds which resources are constrained using node-level capacity and availability data, and then chooses which pods to take action on based on the pod's usage of the constrained resource.  If neccessary, capacity for resources a pod consumes can still be correlated with node-level resources using this format of stats.
+
+```go
+// CoreStats is a top-level container for holding NodeStats and PodStats.  
+type CoreStats struct {  
+  // Overall node resource stats.  
+  Node NodeResources `json:"node"`  
+  // Per-pod usage stats.  
+  Pods []PodUsage `json:"pods"`  
+}  
+
+// NodeStats holds node-level stats.  NodeStats contains capacity and availibility for Node Resources.  
+type NodeResources struct {  
+  // The filesystem device used by node k8s components.  
+  // +optional  
+  KubeletFsDevice string `json:"kubeletfs"`  
+  // The filesystem device used by node runtime components.  
+  // +optional  
+  RuntimeFsDevice string `json:"runtimefs"`  
+  // Stats pertaining to cpu resources.  
+  // +optional  
+  CPU *CpuResources `json:"cpu,omitempty"`  
+  // Stats pertaining to memory (RAM) resources.  
+  // +optional  
+  Memory *MemoryResources `json:"memory,omitempty"`  
+  // Stats pertaining to node filesystem resources.  
+  // +optional  
+  Filesystems []DiskResources `json:"filesystems, omitempty" patchStrategy:"merge" patchMergeKey:"device"`  
```

Comment:
nit: I'd avoid calling this `Disk` since we could hold metrics for `SSD` too in here.